File: CloneA2048/GameBoardView.m
//
//  GameBoard.m
//  CloneA2048
//
//  Created by Mattias Jähnke on 2014-05-26.
//  Copyright (c) 2014 Nearedge. All rights reserved.
//

#import "GameBoardView.h"
#import "GameTileView.h"

#define CONTENT_INSET 10
#define TILE_CORNER_RADIUS 5

@implementation GameBoardView {
    NSMutableArray *_tiles, *_placeholderLayers;
    NSDictionary *_colorScheme;
}

#pragma mark - Init

- (void)_sharedInit {
    _tiles = [NSMutableArray array];
    _placeholderLayers = [NSMutableArray array];

    _colorScheme = [NSJSONSerialization JSONObjectWithData:[NSData dataWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"default-color" ofType:@"json"]] options:0 error:nil];
}

- (id)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self _sharedInit];
    }
    return self;
}

- (id)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self _sharedInit];
    }
    return self;
}

#pragma mark - Public

- (void)setSize:(NSUInteger)size {
    _size = size;
    [self _updatePlaceholderLayers];
}

- (GameTileView *)spawnTileAtPosition:(CGPoint)position {
    GameTileView *tile = [GameTileView new];
    [_tiles addObject:tile];
    [self addSubview:_tiles.lastObject];
    
    tile.colorScheme = _colorScheme;
    tile.position = position;
    tile.cornerRadius = TILE_CORNER_RADIUS;
    tile.frame = [self _frameForPosition:position];
    tile.transform = CGAffineTransformConcat(CGAffineTransformMakeScale(0.1, 0.1f), CGAffineTransformMakeRotation(3.14));
    
    [UIView animateWithDuration:0.3f animations:^{
        ((GameTileView*)_tiles.lastObject).alpha = 1;
        ((GameTileView*)_tiles.lastObject).transform = CGAffineTransformIdentity;
    }];
    
    return tile;
}

- (void)moveTileAtPosition:(CGPoint)fromPosition toPosition:(CGPoint)toPosition {
    [self _tileAtPosition:fromPosition].position = toPosition;
}

- (void)moveAndRemoveTileAtPosition:(CGPoint)fromPosition toPosition:(CGPoint)toPosition {
    GameTileView *tile = [self _tileAtPosition:fromPosition];
    GameTileView *toTile = [self _tileAtPosition:toPosition];
    tile.destory = YES;
    [self moveTileAtPosition:fromPosition toPosition:toPosition];
    
    // "Pop" the merged tile
    assert(toTile);
    [UIView animateWithDuration:.1f animations:^{
        toTile.transform = CGAffineTransformMakeScale(1.2, 1.2);
    } completion:^(BOOL finished) {
        [UIView animateWithDuration:.1f animations:^{
            toTile.transform = CGAffineTransformIdentity;
        }];
    }];
   
}

- (void)animateTiles {
    NSMutableArray *destroyed = [NSMutableArray array];
    for (GameTileView *tile in _tiles) {
        [UIView animateWithDuration:0.1f animations:^{
            CGRect dest = [self _frameForPosition:tile.position];
            tile.bounds = CGRectMake(0,0,dest.size.width, dest.size.height);
            tile.layer.position = CGPointMake(dest.origin.x + tile.bounds.size.width / 2, dest.origin.y + tile.bounds.size.height / 2);
            if (tile.destory) {
                tile.alpha = 0;
            }
        } completion:^(BOOL finished) {
            if (tile.destory) {
                [tile removeFromSuperview];
                [_tiles removeObject:tile];
                [destroyed addObject:tile];
            }
        }];
    }
    [_tiles removeObjectsInArray:destroyed];
}

- (void)updateValuesWithValueArray:(NSArray *)valueArray canSpawn:(BOOL)canSpawn {
    for (int i = 0; i < valueArray.count; i++) {
        for (int j = 0; j < [valueArray[i] count]; j++) {
            GameTileView *tile = [self _tileAtPosition:CGPointMake(i, j)];
            if (canSpawn && [valueArray[i][j] unsignedIntegerValue] > 0 && !tile) {
                tile = [self spawnTileAtPosition:CGPointMake(i, j)];
            }
            if (canSpawn && [valueArray[i][j] unsignedIntegerValue] == 0 && tile) {
                [_tiles removeObject:tile];
                [tile removeFromSuperview];
            }
            assert(!(tile == nil && [valueArray[i][j] unsignedIntegerValue] > 0));
            [tile setValue:[valueArray[i][j] unsignedIntegerValue]];
        }
    }
}

#pragma mark - Internal

- (void)layoutSubviews {
    [super layoutSubviews];
    [self _updatePlaceholderLayers];
    [self animateTiles];
}

- (CGRect)_frameForPosition:(CGPoint)position {
    CGFloat min = MIN(self.frame.size.width, self.frame.size.height) - CONTENT_INSET * 2;
    CGFloat s = roundf(min / _size);
    return CGRectMake(position.y * s + CONTENT_INSET, position.x * s + CONTENT_INSET, s, s);
}

- (GameTileView *)_tileAtPosition:(CGPoint)position {
    for (GameTileView *tile in _tiles) {
        if (CGPointEqualToPoint(position, tile.position) && !tile.destory) {
            return tile;
        }
    }
    return nil;
}

- (void)_updatePlaceholderLayers {
    while (_placeholderLayers.count != _size * _size) {
        if (_placeholderLayers.count < _size * _size) {
            [_placeholderLayers addObject:[CALayer new]];
            [self.layer addSublayer:_placeholderLayers.lastObject];
        } else {
            [_placeholderLayers.lastObject removeFromSuperlayer];
            [_placeholderLayers removeLastObject];
        }
    }
    for (int i = 0; i < _size; i++) {
        for (int j = 0; j < _size; j++) {
            CALayer *layer = (CALayer*)_placeholderLayers[_size * j + i];
            layer.backgroundColor = [UIColor colorWithRed:204/255.0f green:192/255.0f blue:181/255.0f alpha:1].CGColor;
            layer.cornerRadius = TILE_CORNER_RADIUS;
            layer.anchorPoint = CGPointMake(0, 0);
            CGRect f = CGRectInset([self _frameForPosition:CGPointMake(i, j)], 5, 5);
            layer.position = f.origin;
            layer.bounds = f;
        }
    }
}

@end

File: Swift2048/GameBoardView.swift
//
//  GameBoardView.swift
//  CloneA2048
//
//  Created by Mattias Jähnke on 2015-06-30.
//  Copyright © 2015 Nearedge. All rights reserved.
//

import UIKit

typealias BoardPosition = (x: Int, y: Int)

class GameBoardView: UIView {
    let contentInset: CGFloat = 10
    var size = 4 {
        didSet {
            updatePlaceholderLayers()
        }
    }
    var tiles = [GameTileView]()
    var placeholderLayers = [CALayer]()
    var colorScheme: [String : [String : String]] = {
        let data = try! Data(contentsOf: URL(fileURLWithPath: Bundle.main.path(forResource: "default-color", ofType: "json")!))
        do {
            let json = try JSONSerialization.jsonObject(with: data, options: JSONSerialization.ReadingOptions(rawValue: 0)) as! [String : [String : String]]
            return json
        } catch {
            return [String : [String : String]]()
        }
    }()
    
    override func layoutSubviews() {
        super.layoutSubviews()
        updatePlaceholderLayers()
        animateTiles()
    }
    
    func spawnTile(at position: BoardPosition) -> GameTileView {
        let tile = GameTileView(frame: frame(at: position))
        tiles.append(tile)
        addSubview(tile)
        tile.colorScheme = colorScheme
        tile.position = position
        tile.cornerRadius = 5
        tile.transform = CGAffineTransform(scaleX: 0.1, y: 0.1).concatenating(CGAffineTransform(rotationAngle: 3.14))
        
        UIView.animate(withDuration: 0.3) {
            tile.alpha = 1
            tile.transform = CGAffineTransform.identity
        }
        
        return tile
    }
    
    func moveTile(from: BoardPosition, to: BoardPosition) {
        tile(at: from)?.position = to
    }
    
    func moveAndRemoveTile(from: BoardPosition, to: BoardPosition) {
        if let fromTile = tile(at: from), let toTile = tile(at: to) {
            fromTile.destroy = true
            moveTile(from: from, to: to)
            UIView.animate(withDuration: 0.1, animations: { 
                toTile.transform = CGAffineTransform(scaleX: 1.2, y: 1.2)
                }, completion: { _ in
                    UIView.animate(withDuration: 0.1) {
                        toTile.transform = CGAffineTransform.identity
                    }
            })
        }
    }
    
    func animateTiles() {
        var destroyed = [GameTileView]()
        for tile in tiles {
            UIView.animate(withDuration: 0.1, animations: { 
                let dest = self.frame(at: tile.position)
                tile.bounds = CGRect(x: 0, y: 0, width: dest.width, height: dest.height)
                tile.layer.position = CGPoint(x: dest.origin.x + dest.width / 2, y: dest.origin.y + dest.height / 2)
                tile.alpha = tile.destroy ? 0 : 1
                }, completion: { _ in
                    if tile.destroy {
                        tile.removeFromSuperview()
                        destroyed.append(tile)
                    }
            })
        }
        tiles = tiles.filter({ tile -> Bool in
            return !destroyed.contains(tile)
        })
    }
    
    func updateValuesWithModel(_ model: [Int], canSpawn: Bool) {
        for y in 0..<model.size {
            for x in 0..<model.size {
                var t = tile(at: (x: x, y: y))
                if canSpawn && model[x, y] > 0 && t == nil {
                    t = spawnTile(at: (x: x, y: y))
                }
                if canSpawn && model[x, y] == 0 && t != nil {
                    tiles.remove(at: tiles.index(of: t!)!)
                    t!.removeFromSuperview()
                }
                assert(!(t == nil && model[x, y] > 0))
                if let tile = t {
                    tile.value = model[x, y]
                }
            }
        }
    }
    
    fileprivate func updatePlaceholderLayers() {
        while placeholderLayers.count != size * size {
            if placeholderLayers.count < size * size {
                placeholderLayers.append(CALayer())
                layer.addSublayer(placeholderLayers.last!)
            } else {
                placeholderLayers.last!.removeFromSuperlayer()
                placeholderLayers.removeLast()
            }
        }
        
        for y in 0..<size {
            for x in 0..<size {
                let layer = placeholderLayers[size * x + y]
                layer.backgroundColor = UIColor(red: 204.0/255.0, green: 192.0/255.0, blue: 181.0/255.0, alpha: 1).cgColor
                layer.cornerRadius = 5
                layer.anchorPoint = CGPoint(x: 0, y: 0)
                let rect = frame(at: (x: x, y: y)).insetBy(dx: 5, dy: 5)
                layer.position = rect.origin
                layer.bounds = rect
            }
        }
    }
    
    fileprivate func tile(at position: BoardPosition) -> GameTileView? {
        return tiles.filter { $0.position == position && !$0.destroy }.first
    }
    
    fileprivate func frame(at position: BoardPosition) -> CGRect {
        let minSize = min(frame.size.width, frame.size.height) - contentInset * 2
        let s = round(minSize / CGFloat(size))
        return CGRect(x: CGFloat(position.x) * s + contentInset, y: CGFloat(position.y) * s + contentInset, width: s, height: s)
    }
    
}
